// Protos for reachability analysis.

syntax = "proto2";

package phd.deeplearning.ml4pl;

// A collection of control flow graphs.
message ControlFlowGraphSet {
  repeated ControlFlowGraph graph = 1;
  optional int64 initial_random_seed = 2;
}

// A control flow graph is a collection of basic blocks and edges between them.
message ControlFlowGraph {
  optional string name = 1;
  // The nodes in the CFG.
  repeated BasicBlock block = 2;
  // The connections between nodes of the CFG.
  repeated ControlFlowGraphEdge edge = 3;
  optional int64 entry_block_index = 4;
  repeated int64 exit_block_index = 5;
}

// A basic block.
message BasicBlock {
  required string name = 1;
  optional string text = 2;
}

// A CFG edge is a pair of indices into a ControlFlowGraph.block list that
// identifies the source and destination BasicBlocks.
message ControlFlowGraphEdge {
  required int64 src_index = 1;
  required int64 dst_index = 2;
}

// The result of attempting to extract LLVM bytecode from an input source.
message LlvmBytecode {
  required string source_name = 1;
  required string relpath = 2;
  required string lang = 3;
  required string cflags = 4;
  required string bytecode = 5;
  required int32 clang_returncode = 6;
  required string error_message = 7;
}

message ControlFlowGraphFromLlvmBytecode {
  required int64 bytecode_id = 1;
  // Bytecodes can contain multiple control-flow graphs. This is the ID of the
  // control flow graph, an integer series 0..n.
  required int64 cfg_id = 2;
  required ControlFlowGraph control_flow_graph = 3;
  required int32 status = 4;
  required string error_message = 5;
  required int64 block_count = 6;
  required int64 edge_count = 7;
  required bool is_strict_valid = 8;
}
